src/Adapter/MSTest.TestAdapter/Execution/TestMethodRunner.cs

```diff
@@ -81,63 +148,159 @@ internal UnitTestResult[] Execute()
 
         UnitTestResult[]? result = null;
 
+        bool shouldRunClassCleanup = false;
+        bool shouldRunClassAndAssemblyCleanup = false;
+        bool runFailed = false;
+        LogMessageListener logListener = new(_captureDebugTraces);
+        try
+        {
+            try
+            {
+                // Run the assembly Initialize methods if required.
+                // Assembly initialize can throw exceptions in which case we need to ensure that we fail the test.
+                _testMethodInfo.Parent.Parent.RunAssemblyInitialize(_testContext.Context);
+            }
+            finally
+            {
+                initializationLogs = logListener.GetAndClearStandardOutput();
+                initializationTrace = logListener.GetAndClearDebugTrace();
+                initializationErrorLogs = logListener.GetAndClearStandardError();
+                initializationTestContextMessages = _testContext.GetAndClearDiagnosticMessages();
+            }
+        }
+        catch (Exception ex)
+        {
+            runFailed = true;
+            RunCatch(ex, ref result);
+            RunFinally(ref result, ref initializationLogs, ref initializationTrace, ref initializationErrorLogs, ref initializationTestContextMessages);
+        }
+
+        Type attributeType = _testMethodInfo.Parent.ClassAttribute.GetType();
+        if ((_testMethodInfo.Parent.ClassAttribute is STATestClassAttribute || attributeType.IsSubclassOf(typeof(STATestClassAttribute)))
```

Comment:
As a rule of thumb, always try to have simpler checks done first (e.g. `runFailed` is less expensive than others so should be checked first).